src/util/memoizer.ts

```diff
@@ -0,0 +1,445 @@
+    });
+  };
+
+  public toImmutableObjectKey = (js: any): ImmutableObjectKey => {
+    // This is necessary because of the oddities of CQL
+    // It allows ignoring non-set values in tuples to be compared correctly with set as null values in tuples
+    if (js === null || js === undefined) return null;
+
+    // Handle the edge case of functions
+    if (typeof js === 'function')
+      return Immutable.Map({
+        name: js.toString(),
+        __instance: js.constructor.name
+      });
+
+    // Simple return non-objects
+    if (typeof js !== 'object') return js;
+
+    // Handle objects - normalize as necessary to generate unique keys
+    switch (js.constructor) {
+      case Array:
+        return Immutable.Seq(js)
+          .map((x: any) => this.toImmutableObjectKey(x))
+          .toList();
+
+      case Code:
+        return Immutable.Map({
+          code: this.toImmutableObjectKey(js.code),
+          system: this.toImmutableObjectKey(js.system),
+          __instance: js.constructor.name
+        });
+
+      case Date:
+        return Immutable.Map({
+          epochMs: js.getTime(),
+          __instance: js.constructor.name
+        });
+
+      case DateTime:
+        if (typeof js.timezoneOffset === 'number' && js.timezoneOffset !== 0)
+          return Immutable.Seq(js.convertToTimezoneOffset(0))
+            .map((x: any) => this.toImmutableObjectKey(x))
+            .toMap()
+            .set('__instance', js.constructor.name);
+        else
+          return Immutable.Seq(js)
+            .map((x: any) => this.toImmutableObjectKey(x))
+            .toMap()
+            .set('__instance', js.constructor.name);
+
+      case Interval:
+        return Immutable.Seq(js.toClosed())
+          .map((x: any) => this.toImmutableObjectKey(x))
+          .toMap()
+          .set('__instance', js.constructor.name);
+
+      case Quantity:
+        if (!js.unit) {
+          return Immutable.Map({
+            value: js.value ?? null,
+            unit: null,
+            __instance: js.constructor.name
+          });
+        }
+
+        // Get the normalized base unit
+        const baseUnitKey = ucumUtilInstance.commensurablesList(js.unit)[0];
+
+        if (!baseUnitKey) {
+          // No units found - normalization not possible and use provided values
+          return Immutable.Map({
+            value: js.value ?? null,
+            unit: js.unit ?? null,
+            __instance: js.constructor.name
+          });
+        } else {
+          // Unit was found - convert to baseUnit and normalize
+          const baseUnitKeyCode = baseUnitKey[0].csCode_;
+          const conversionValue = convertUnit(js.value, js.unit, baseUnitKeyCode);
+          const finalValue = conversionValue ? decimalAdjust('round', conversionValue, -8) : null;
+          return Immutable.Map({
+            value: finalValue ?? null,
+            unit: baseUnitKeyCode ?? null,
+            __instance: js.constructor.name
+          });
```

Comment:
Two questions:

1. Is this for efficiency, or is this part of how we get `Immutable.is` to work like our `equal` implementation?
2. Does this normalize Quantity-type arguments in the memoized function call as well?  In other words, if I memoized a function like `toString(q: Quantity)` and called `toString(new Quantity(1, 'm'))` (which returns `"1 'm'"`), and then I called `toString(new Quantity(100, 'cm'))` -- would the arguments get normalized to the same key and result in the second call also returning `"1 'm'"`?

Reply:
1. It's necessary to match the equals implementation.  

In quantity.ts, the `equals` implementation has two quantities - and so it converts one quantity to match the units of the other for comparison.

In memoizer.ts, there is only one quantity for which a key is being generated.  So "normalized units" have to be chosen so that all quantities are comparable.  If the quantity.ts implementation converted both quantities to a normalized value for comparison, then the implementations would be exactly the same.

2. It depends on the choice of memoizer implementation.  I'll give two examples - one is a normalized toString, the other non-normalized:

Example 1: Normalized toString

![image](https://user-images.githubusercontent.com/109760379/194118749-8ca55374-454d-4991-94e2-4a3394b1d982.png)

Example 2: Non-normalized toString

![image](https://user-images.githubusercontent.com/109760379/194118915-692a1aed-ded1-42d5-a8b9-83736ed8e8c8.png)

Reply:
> Here's one more just for fun 

Weren't the other ones fun too?

Yeah, OK.  So it's as I suspected, but... now that I think about it more, I realize that for the functions we are memoizing, it shouldn't actually matter.  I also see (and tested) that when it drops back to the original (non-Immutable) objects, it maintains the original units (and doesn't force the normalized units).  So `distinct {1 'm', 200 'cm', 4000 'mm'}` comes back exactly as I passed it in (not in normalized units).  So I think we're good!